lightningd/pay.c

```diff
@@ -890,6 +890,98 @@ param_route_hop(struct command *cmd, const char *name, const char *buffer,
 	return NULL;
 }
 
+static struct command_result *json_sendonion(struct command *cmd,
+					     const char *buffer,
+					     const jsmntok_t *obj UNNEEDED,
+					     const jsmntok_t *params)
+{
+	u8 *onion;
+	struct onionpacket *packet;
+	enum onion_type failcode;
+	struct htlc_out *hout;
+	struct route_hop *first_hop;
+	struct sha256 *payment_hash;
+	struct channel *channel;
+	struct lightningd *ld = cmd->ld;
+	struct wallet_payment *payment;
+	const char *label;
+
+	if (!params || !deprecated_apis) {
+		if (!param(cmd, buffer, params,
+			   p_req("onion", param_bin_from_hex, &onion),
+			   p_req("first_hop", param_route_hop, &first_hop),
+			   p_req("payment_hash", param_sha256, &payment_hash),
+			   p_opt("label", param_escaped_string, &label),
+			   NULL))
+			return command_param_failed();
+	}
+
+	packet = parse_onionpacket(cmd, onion, tal_bytelen(onion), &failcode);
+
+	if (!packet)
+		return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+				    "Could not parse the onion. Parsing failed "
+				    "with failcode=%d",
+				    failcode);
+
+	/* FIXME if the user specified a short_channel_id, but no peer nodeid,
+	 * we need to resolve that first. */
+
+	channel = active_channel_by_id(ld, &first_hop->nodeid, NULL);
+	if (!channel) {
+		struct json_stream *data
+			= json_stream_fail(cmd, PAY_TRY_OTHER_ROUTE,
+					   "No connection to first "
+					   "peer found");
+
+		json_add_routefail_info(data, 0, WIRE_UNKNOWN_NEXT_PEER,
+					&ld->id, &first_hop->channel_id,
+					node_id_idx(&ld->id, &first_hop->nodeid),
+					NULL);
+		json_object_end(data);
+		return command_failed(cmd, data);
+	}
+
+	hout = send_onion(cmd->ld, packet, first_hop, payment_hash, channel,
+			  &failcode);
+
+	payment = tal(hout, struct wallet_payment);
+	payment->id = 0;
+	payment->payment_hash = *payment_hash;
+		payment->status = PAYMENT_PENDING;
+	payment->msatoshi = first_hop->amount;
```

Comment:
This should be the last_hop amount I think.

Reply:
We don't have access to that though, since we're given a fully wrapped onion only. We don't know what happens to the funds, apart from how much is sent initially.

Reply:
`NULL` in the DB and 0 in-memory sounds like the correct answer here, since we have no idea how much will arrive at the destination and we still have the `msatoshi_sent` amount here. Does that sound ok to you @rustyrussell and @niftynei?

Reply:
An improvement would be to make `msatoshi` in `struct wallet_payment` a pointer since it can be `NULL`, that would come at the cost of an extra allocation, but be more correct?